File: back-end/src/main/java/com/huyphan/mappers/SliceMapper.java
package com.huyphan.mappers;

import com.huyphan.dtos.SliceDto;
import java.util.List;
import java.util.stream.Collectors;
import org.modelmapper.Converter;
import org.modelmapper.ModelMapper;
import org.modelmapper.TypeToken;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Slice;
import org.springframework.stereotype.Component;

@Component
public class SliceMapper<TDto, TDomain> {

    @Autowired
    private ModelMapper modelMapper;

    public SliceDto<TDto> toDto(Slice<TDomain> data, ToDtoMapper<TDto, TDomain> toDtoMapper) {
        Class<SliceDto<TDto>> desType = new TypeToken<SliceDto<TDto>>() {
        }.getRawType();
        Converter<List<TDomain>, List<TDto>> converter = (context) -> {
            List<TDomain> domainList = context.getSource();

            return domainList.stream().map(toDtoMapper::toDto).collect(Collectors.toList());
        };
        modelMapper.createTypeMap(data.getClass(), desType).addMappings(
                mapper -> mapper.using(converter)
                        .map(Slice<TDomain>::getContent, SliceDto<TDto>::setContent)
        );

        return modelMapper.map(data, desType);
    }
}

File: client/src/services/mappers/slice-mapper.ts
import Slice from '../../models/slice';
import SliceDto from '../dtos/slice-dto';
import { MapperFromDtoFunc } from './mapper';

export namespace SliceMapper {
  export function fromDto<T, TDto>(
    data: SliceDto<T>,
    mapfromDto: MapperFromDtoFunc<T, TDto>,
  ): Slice<TDto> {
    return {
      isLast: data.last,
      page: data.number,
      size: data.size,
      content: data.content.map((value) => mapfromDto(value)),
    };
  }
}
